google-cloud-firestore/src/main/java/com/google/cloud/firestore/Watch.java

```diff
@@ -0,0 +1,580 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.firestore;
+
+import com.google.api.gax.grpc.GrpcStatusCode;
+import com.google.api.gax.rpc.ApiException;
+import com.google.api.gax.rpc.ApiStreamObserver;
+import com.google.cloud.firestore.DocumentChange.Type;
+import com.google.common.base.Preconditions;
+import com.google.firestore.v1beta1.Document;
+import com.google.firestore.v1beta1.ListenRequest;
+import com.google.firestore.v1beta1.ListenResponse;
+import com.google.firestore.v1beta1.Target;
+import com.google.firestore.v1beta1.Target.QueryTarget;
+import com.google.firestore.v1beta1.TargetChange;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Timestamp;
+import io.grpc.Status;
+import io.grpc.Status.Code;
+import io.grpc.StatusException;
+import io.grpc.StatusRuntimeException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nullable;
+import org.threeten.bp.Instant;
+
+/**
+ * Watch provides listen functionality and exposes snapshot listeners. It can be used with any
+ * valid Firestore Listen target.
+ *
+ * This class is thread-compatible when called through the methods defined in ApiStreamObserver.
+ * It synchronizes on its own instance so it is advisable not to use this class for external
+ * synchronization.
+ */
+class Watch implements ApiStreamObserver<ListenResponse> {
+  /**
+   * Target ID used by watch. Watch uses a fixed target id since we only support one target per
+   * stream.
+   */
+  private static final int WATCH_TARGET_ID = 0xD0;
```

Comment:
Does the value 0xD0 actually mean anything?

Reply:
No, I just chose it to be different from Node Watch. Maybe one day it will help us debug something on the backend. 

Added comment.